README.md
# iOS OTP Authentication

This demo is of iOS based OTP Authentication, used to verify your mobile number with OTP (One Time Password).

## Table of contents
- [iOS Support](#iOS-support)
- [Demo](#demo)
- [Features](#features)
- [Getting started](#getting-started)
- [Usage](#usage)
- [Methods](#methods)
- [Directive options](#directive-options)
- [Want to Contribute?](#want-to-contribute)
- [Need Help / Support?](#need-help)
- [Collection of Components](#collection-of-Components)
- [Changelog](#changelog)
- [License](#license)
- [Keywords](#Keywords)


## iOS Support

Version - iOS 13.x or later

We have tested our program in above versions, however you can use it in other versions as well


## Demo
[![](OTP_Authentication.png)](https://github.com/weblineindia/iOS-OTP-Authentication/blob/master/OTP_Authentication.png)

------

## Features
* Select country with flag & country code
* Verify mobile number with OTP all over the world


## Getting started
Download this sample project and import swift utility resource files in your project. 

## Usage
Process setup is described below to integrate in sample project.

Configure your required property of  countriesPickerVcObj
Also set delegate and implement method for call back selected country details
e.g. 

    extension ViewController: CountriesViewControllerDelegate {   
    func countriesViewController(_ countriesViewController: CountriesViewController, didSelectCountry country: Country) {
        //Return last selected country details
        }}

### Create  CountryModel object for store last selected country details
    var selectedCountry:CountryModel?


## Methods

### PinView Class For OTP(One Time Password)
- Setup pinview and config validation you want to setup
below field config 
1. Create PinView Outlet of StackView
2. Access config form Pinview outlet properties
setup below validation base on your requirement
---------------------------------------------------------------------------------------------------------
        var config:PinConfig!     = PinConfig()
        // Length of OTP
        config.otpLength          = .six
        // Secure text mask color
        config.dotColor           = .black
        config.lineColor          = #colorLiteral(red: 0.8265652657, green: 0.8502194881, blue: 0.9000532627, alpha: 1)
        //Space beetween two text field
        config.spacing            = 30
        //set secure text
        config.isSecureTextEntry  = true
        //show place hodder
        config.showPlaceHolder    = true
---------------------------------------------------------------------------------------------------------

### Setup text field delegate
 like Enter only specify values in textfield
   
    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        ///- For Phone number validation
        if textField == txtPhoneNumberField {
            let aSet = NSCharacterSet(charactersIn:"0123456789").inverted
            let compSepByCharInSet = string.components(separatedBy: aSet)
            let numberFiltered = compSepByCharInSet.joined(separator: "")
            return string == numberFiltered
        }
        else{
            return true
        }
    }
    
### setup OTP view
    - setUpOTPView: this method provide setup of OTP text field like below


### validateOtpAndVerifyViaServer
validate number with server use of below method
- enter OTP 
- callback response
    
      self.countriesPickerVcObj.verifyOTP(otpCode: "123456", onSuccess:(authDataResult,bool) , onFailure:(error,isError)


 ### send OTP to given details
 - enter country code
 - enter phone numner
 - call back validate details information
 
       self.countriesPickerVcObj.validateContryAndPhoneNumber(contryCode: "+1", phoneNumber: "9XXXXXXXXX", isValidDetails:(isValid,strMsg)


### validatePhoneNumber
- Param 
  - number
- callBack
  - return validationId & error details
this method validate phone number and send OTP


### verifyOTPAndSigIn
Param
 - enter OTP
callback
 - AuthDataResult object & error details

### sendOtpAfterValidatePhonenumber
Param
 - view: for show indicator
 - phoneNowithCountryCode : for Example "+919XXXXXXXX"
 sucesss
  - return isSendSucessOTP status of send otp or not 
  - return errorMessage show error
call this method after check text field details


### verifyOTP
 Param
  - otpCoed
OnSuccess
 - AuthDataResult : get user data
 - isVarifyOTP : check otp verify or not
onFailure
 - error : display error
 - isVarifyOTP : check otp verify or not

------

## Directive options

### Firebase project setup steps
Create  Project setup in firebase console using below URL
https://console.firebase.google.com/
 Please Enable Phone Number sign-in for your Firebase project in console    
https://firebase.google.com/docs/auth/ios/phone-auth

Download GoogleService-Info.Plist file and add into project 
install firebase pod and run project 


### Pod installation process
The `OTPAuthentication` demo use in you app then same option as the `<OTPAuthentication>`
install firebase pod
- pod 'Firebase/Analytics'
- pod 'Firebase/Auth'
   
show country picker use of below method help of CountriesViewController class object
- CountriesViewController.show(countriesViewController: self.countriesPickerVcObj, toVar: self)
 
- select country and Phone number

- Press send button to validate entered details and send OTP help of below mehods

      FirebaseAuthenticate.firSharedInstance.sendOtpAfterValidatePhonenumber

- Verify OTP with below method

      FirebaseAuthenticate.firSharedInstance.verifyOTP
      
------

## Want to Contribute?

- Created something awesome, made this code better, added some functionality, or whatever (this is the hardest part).
- [Fork it](http://help.github.com/forking/).
- Create new branch to contribute your changes.
- Commit all your changes to your branch.
- Submit a [pull request](http://help.github.com/pull-requests/).

------

## Need Help? 
We also provide a free, basic support for all users who want to use this OTP verification demo in project. In case you want to customize this OTP demo input to suit your development needs, then feel free to contact our [iOS
developers](https://www.weblineindia.com/hire-ios-app-developers.html).

 ------
 
 ## Collection of Components
 We have built many other components and free resources for software development in various programming languages. Kindly click here to view our [Free Resources for Software Development.](https://www.weblineindia.com/software-development-resources.html)
 
------

## Changelog
Detailed changes for each release are documented in [CHANGELOG.md](./CHANGELOG.md).

## License

[MIT](LICENSE)

[mit]: https://github.com/weblineindia/iOS-OTP-Authentication/blob/master/LICENSE

## Keywords
OTP, Firebase Authentication, One Time Password, OTP For iOS, OTP Authentication, Mobile Number Verficiation,
OTP Verification, Firebase OTP Authentication, weblineindia


CHANGELOG.md
1.0.1 (July 03, 2020)
Initial release - Sample source code for OTP Authentication demo.

1.1 (Nov 26, 2021)
Latest OS compatibility changes.


DemoLoginWithOTP/ViewController.swift
//
//  ViewController.swift
//  DemoLoginWithOTP
//
//  Created by WeblineIndia on 23/12/19.
//  Copyright © 2019 WeblineIndia. All rights reserved.
//

import UIKit
import FirebaseAuth

class ViewController: UIViewController {
    @IBOutlet weak var imgview: UIImageView!
    
    @IBOutlet weak var lblFlag: UITextField!
    @IBOutlet weak var viewLogin: UIView!
    
    @IBOutlet weak var viewNumber: UIView!
    @IBOutlet weak var viewOtp: UIView!
    @IBOutlet weak var pinView : PinView!
    @IBOutlet weak var btnSendOTP : UIButton!
    
    @IBOutlet weak var lblTitle: UILabel!
    @IBOutlet weak var lblSubTitle: UILabel!

    @IBOutlet weak var txtPhoneNumber: UITextField!
    @IBOutlet weak var lblCountryCode: UILabel!
    var countriesPickerVcObj = CountriesViewController()
    var selectedCountry:CountryModel?
    
    //MARK: View Life Cycle Method
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        let tap: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(dismissKeyBord))
        view.addGestureRecognizer(tap)
        
        viewLogin.layer.cornerRadius = 10
        viewLogin.clipsToBounds  = true
        btnSendOTP.layer.cornerRadius = btnSendOTP.frame.size.height / 2
        btnSendOTP.clipsToBounds = true
        
        viewNumber.layer.cornerRadius =  viewNumber.frame.size.height / 2
        viewNumber.clipsToBounds  = true
        viewNumber.layer.borderColor = UIColor.blue.cgColor
        viewNumber.layer.borderWidth = 0.2
        
        setupCountryPicker()
        setUpOTPView()
        
         NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow), name: UIResponder.keyboardWillShowNotification, object: nil)
          NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillHide), name: UIResponder.keyboardWillHideNotification, object: nil)

        
    }
    
    @objc func keyboardWillShow(notification: NSNotification) {
        if let keyboardSize = (notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? NSValue)?.cgRectValue {
            if self.view.frame.origin.y == 0 {
                self.view.frame.origin.y -= keyboardSize.height
            }
        }
    }

    @objc func keyboardWillHide(notification: NSNotification) {
        if self.view.frame.origin.y != 0 {
            self.view.frame.origin.y = 0
        }
    }
    
    override func viewDidLayoutSubviews() {
        viewLogin.layer.masksToBounds = false
        viewLogin.layer.shadowColor = UIColor.gray.cgColor
        viewLogin.layer.shadowOffset = CGSize(width: 0, height: 5)
        viewLogin.layer.shadowOpacity = 0.35
        viewLogin.layer.shadowPath = UIBezierPath(roundedRect: viewLogin.bounds, cornerRadius: viewLogin.layer.cornerRadius).cgPath
    }
    
    @objc func dismissKeyBord(){
        self.view.endEditing(true)
    }
     ///- setup for country code list
    func setupCountryPicker(){
        self.countriesPickerVcObj = CountriesViewController()
        self.countriesPickerVcObj.delegate = self
        if let info = countriesPickerVcObj.getCountryAndName() {
            selectedCountry = info
            self.lblFlag.text = info.flag!
            self.lblCountryCode.text = info.countryCode!
        }
    }
    
     ///- Config OTP view
    func setUpOTPView(){
       var config:PinConfig!     = PinConfig()
         config.otpLength          = .six
         config.dotColor           = .black
         config.lineColor          = #colorLiteral(red: 0.8265652657, green: 0.8502194881, blue: 0.9000532627, alpha: 1)
         config.spacing            = 30
         config.isSecureTextEntry  = true
         config.showPlaceHolder    = true
        
        pinView.config = config
        pinView.setUpView()
        //Delegate for validate text
        txtPhoneNumber.delegate = self
        txtPhoneNumber.becomeFirstResponder()
      
    }
   
    
    ///- Enter only specify code in text fields
    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
    
        ///- For Phone number validation
        if textField == txtPhoneNumber {
            let aSet = NSCharacterSet(charactersIn:"0123456789").inverted
            let compSepByCharInSet = string.components(separatedBy: aSet)
            let numberFiltered = compSepByCharInSet.joined(separator: "")
            return string == numberFiltered
        }
        else{
            return true
        }
    }
   
    ///-  submit OTP to server
    func validateOtpAndVerifyViaServer(){
        var otpCode:String = ""
        do {
            otpCode = try self.pinView.getOTP()
            print("Enter OTP is:\(otpCode)")
            HelperClass.showActivityIndicatory(uiView: self.view)
            FirebaseAuthenticate.firSharedInstance.verifyOTP(otpCode: otpCode, onSuccess: { (useraData, isVerified) in
               
                HelperClass.hideActivityIndicator(uiView: self.view)
                self.showAlertMessage("OTP Verified Successfully")

                self.lblTitle.text = "Login"
                self.lblSubTitle.text = "Enter your mobile number to receive a verification code"
                self.btnSendOTP.setTitle("Send OTP", for: .normal)

                self.viewOtp.isHidden = true
                self.viewNumber.isHidden = false

                self.txtPhoneNumber.text = ""
                self.imgview.image = UIImage.init(named: "registration")
                
            }) { (error, isvalid) in
                HelperClass.hideActivityIndicator(uiView: self.view)
                self.showAlertMessage(error.localizedDescription)
            }
        } catch OTPError.inCompleteOTPEntry {
            self.showAlertMessage(OTPError.inCompleteOTPEntry.localizedDescription)
            print("Incomplete OTP Entry valid error")
        } catch let error {
            print(error.localizedDescription)
        }
    }
    
    //MARK: Action Methods
     ///- verify number and Send OTP request
    @IBAction func btnSendOTP(_ sender : UIButton){
        self.dismissKeyBord()
        if btnSendOTP.titleLabel?.text == "Send OTP" {
            ValidationManager.sharedInstance.validateContryAndPhoneNumber(contryCode: selectedCountry?.countryCode! ?? "", phoneNumber: txtPhoneNumber.text!) { (isValidDetails, message) in
                if isValidDetails {
                    //Send OTP to given number
                    FirebaseAuthenticate.firSharedInstance.sendOtpAfterValidatePhonenumber(view: self.view, phoneNowithCountryCode: message) { (isSucess, message) in
                        if isSucess {
                          
                            self.viewOtp.isHidden = false
                            self.viewNumber.isHidden = true
                            self.btnSendOTP.setTitle("Verify", for: .normal)
                            self.lblTitle.text = "Verification"
                            self.imgview.image = UIImage.init(named: "verification")
                            self.lblSubTitle.text = "Enter A 6 digit number that was sent to " + "\(self.selectedCountry?.countryCode ?? "+91")" + "\(self.txtPhoneNumber.text ?? "")"
                            self.pinView.textFields[0].becomeFirstResponder()
                        }else {
                            self.showAlertMessage(message)
                        }
                    }
                } else{
                    self.showAlertMessage(message)
                }
            }
        }else{
            validateOtpAndVerifyViaServer()
        }
    }
    
    /// - Resend OTP
    @IBAction func btnResend(_ sender : UIButton){
        self.dismissKeyBord()
        pinView.clearOTP()
        ValidationManager.sharedInstance.validateContryAndPhoneNumber(contryCode: selectedCountry?.countryCode! ?? "", phoneNumber: txtPhoneNumber.text!) { (istrue, message) in
            FirebaseAuthenticate.firSharedInstance.sendOtpAfterValidatePhonenumber(view: self.view, phoneNowithCountryCode: message) { (isResend, message) in
                if isResend  {
                    //show alert resend otp sucess
                    }else {
                    //show alert resend otp faild
                }
            }
        }
    }
    
    @IBAction func btnCancelOTP(_ sender: UIButton) {
        self.btnSendOTP.setTitle("Send OTP", for: .normal)
        self.txtPhoneNumber.text = ""
        self.lblSubTitle.text = "Please enter your mobile no with country code to receive OTP."
        self.txtPhoneNumber.becomeFirstResponder()
    }
    @IBAction func btnCountryPicker(_ sender: UIButton) {
        
        DispatchQueue.main.async {
            CountriesViewController.show(countriesViewController: self.countriesPickerVcObj, toVar: self)
        }
    }
    ///- show alert message
       func showAlertMessage(_ message : String){
           let alert = UIAlertController(title: "", message: message, preferredStyle: .alert)
           let ok = UIAlertAction(title: "OK", style: .default) { (action) in
               self.pinView.clearOTP()
           }
           alert.addAction(ok)
           self.present(alert, animated: true, completion: nil)
       }
}

extension ViewController : UITextFieldDelegate {
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        return true
    }
///- open country picker
//    func textFieldShouldBeginEditing(_ textField: UITextField) -> Bool {
//        if textField == self.txtCountryField {
//            DispatchQueue.main.async {
//                CountriesViewController.show(countriesViewController: self.countriesPickerVcObj, toVar: self)
//            }
//            return false
//        }
//        return true
//    }
}

///Country code picker delegate
extension ViewController: CountriesViewControllerDelegate {
   
    func countriesViewController(_ countriesViewController: CountriesViewController, didSelectCountry country: Country) {
        if let info = self.countriesPickerVcObj.getCountryAndName(country.countryCode) {
            selectedCountry = info
            lblCountryCode.text = info.countryCode!
            lblFlag.text = info.flag
        }
    }
}
